### PR TITLE
rpc: Remove mempool global from miner

### PR DESCRIPTION
**Pull request by maflcko**
The miner needs read-only access to the mempool. Instead of using the mutable global `::mempool`, keep a immutable reference to a mempool that is passed to the miner. Apart from the obvious benefits of removing a global and making things immutable, this might also simplify testing with multiple mempools.

### CODE REVIEW

**Comment by practicalswift** [comment]
Strong concept ACK

Also welcome from a fuzz testing perspective :)

**Comment by jnewbery** [comment]
Concept ACK

**Comment by JeremyRubin** [comment]
Concept ACK, but this will make me sad from a rebasing perspective ;)

**Comment by promag** [comment]
Concept ACK.

File: src/bench/bench.h [hunk 1]
```diff
@@ -14,6 +14,9 @@
 #include <boost/preprocessor/cat.hpp>
 #include <boost/preprocessor/stringize.hpp>
 
+struct RegTestingSetup;
```
Comment (emilengler) [inline src/bench/bench.h]: Maybe I oversaw it but is it really necessary to include this variable into the header?
As far as I know it should work just fine if just the pointer is included in the header file.

Comment (sipa) [reply src/bench/bench.h]: This is not declaring a variable, it's forward declaring the type RegTestingSetup.

Comment (emilengler) [reply src/bench/bench.h]: But is it still required?

Comment (promag) [reply src/bench/bench.h]: Yes, it's defined in src/test/util/setup_common.h which isn't included at this point.

File: src/bench/bench.cpp [hunk 2]
```diff
@@ -113,6 +115,7 @@ void benchmark::BenchRunner::RunAll(Printer& printer, uint64_t num_evals, double
 
     for (const auto& p : benchmarks()) {
         RegTestingSetup test{};
+        g_testing_setup = &test;
```
Comment (promag) [inline src/bench/bench.cpp]: nit, `assert(g_testing_setup == nullptr)` before

**Comment by promag** [comment]
Code review ACK fa9c7398087b821ccebf522ee5184eca7fa81445.

**Review by promag — COMMENTED** [review]
Code review ACK fa9c7398087b821ccebf522ee5184eca7fa81445.

**Comment by MarcoFalke** [comment]
Addressed feedback by @promag

**Comment by promag** [comment]
ACK faa92a2297b4a6aebdd58d1818c428f1c0346078.

**Comment by fjahr** [comment]
ACK faa92a2297b4a6aebdd58d1818c428f1c0346078

Reviewed code, tested locally.

File: src/miner.cpp [hunk 3]
```diff
@@ -45,7 +45,9 @@ BlockAssembler::Options::Options() {
     nBlockMaxWeight = DEFAULT_BLOCK_MAX_WEIGHT;
 }
 
-BlockAssembler::BlockAssembler(const CChainParams& params, const Options& options) : chainparams(params)
+BlockAssembler::BlockAssembler(const CTxMemPool& mempool, const CChainParams& params, const Options& options)
+    : chainparams(params),
```
Comment (jnewbery) [inline src/miner.cpp]: nit: order member initialization list the same as the parameter list (and preferably on one line)

File: src/rpc/mining.cpp [hunk 4]
```diff
@@ -179,9 +179,11 @@ static UniValue generatetodescriptor(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, strprintf("Cannot derive script without private keys"));
     }
 
+    const CTxMemPool& mempool = EnsureMemPool();
```
Comment (jnewbery) [inline src/rpc/mining.cpp]: nit: place this either at the start of the function (so it exits early if there is no mempool), or immediately before the call to `generateBlocks()`. Placing it between the code that creates `coinbase_script` and the coinbase_script CHECK seems random.

Same comment for `generatetoaddress()`

**Review by jnewbery — COMMENTED** [review]
Code review ACK faa92a2297b4a6aebdd58d1818c428f1c0346078

A couple of nits inline. Feel free to ignore.

File: src/miner.cpp (hunk 3 above)
Comment (MarcoFalke) [reply src/miner.cpp]: I prefer multiple lines

File: src/rpc/mining.cpp (hunk 4 above)
Comment (MarcoFalke) [reply src/rpc/mining.cpp]: Might do in a follow-up to not invalidate the three reviews

**Pull request merged**